crates/sc-subspace-block-relay/src/consensus.rs

```diff
@@ -453,7 +453,8 @@ where
             FromBlock::Number(n) => BlockId::<Block>::Number(n),
         };
 
-        let mut total_size: usize = 0;
+        // This is a compact length encoding of max number of blocks to be sent in response
```

Comment:
Thanks. Can you explain what is going on here? Also, this is a limit on number of blocks, while we want a limit on both num blocks and response size (in bytes)

Reply:
Vectors in SCALE codec are encoded as compact length prefix + concatenation of encodings of individual elements. Using `Compact::compact_len()` we can calculate how many bytes will compact encoding of the number will take. In this case we do that assuming we will be able to push max number of blocks into the response (essentially the upper bound for the prefix).

This is still a bit lazy, we might have 1 free byte in case of small number of blocks, but at least we will never not exceed the limit.

https://docs.substrate.io/reference/scale-codec/ has all of the details about SCALE encoding.

Reply:
SG. But I would still prefer the solution in 1883, as it avoids extra `encoded_size()` calls + actual encode in the end

The best solution would to be to have a `try_encode_to()` kind of thing in the `Encode` trait, which fails if it exceeds a limit, but that would need upstream changes

Reply:
It is more about number of elements. Large or not the calculation essentially ends up being equivalent to `vec.len()` calls and such.